---
title: "DESeq2_DEA"
author: "James Spurrier"
date: "`r Sys.Date()`"
output: html_document
editor:
  mode: source
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential Expression Analysis with DESeq2

```{r}
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(ComplexHeatmap)
library(PCAtools)
library(vidger)
library(DEGreport)
library(RColorBrewer)
library(circlize)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(genefilter)
library(clusterProfiler)
library(EnhancedVolcano)
library(ggvenn)
library(VennDiagram)
```

```{r}
sample_sheet <- read.csv("metadata/rnaseq/SRP104161/SRP104161_SampleSheet.csv")
#sample_sheet <- read.csv("metadata/rnaseq/metadata_deseq_SRP104161_SRP406132.csv")

resultsDir <- file.path("results/rnaseq/SRP104161/deseq2")
figureDir <- file.path(resultsDir, "figures")
exp <- paste0("786O_vs_M1A")
```

```{r}
metadata <- sample_sheet[1:4, ]
```

```{r}
dir.create(resultsDir)
dir.create(figureDir)
```

```{r}
README <-   'DESeq2 v1.42.1 differential expression analysis
Dataset: SRP104161
Description: Reads were first quantified by RSubread. 
Notes: DEA for 786-O/786-M1A samples, OS-RC2/OS-LM1 samples, plus a batch corrected analysis of all metastatic vs non-metastatic. 
CHANGE: Adjusted low count prefilter to <rowSums( counts(dds, normalized=TRUE) >= 10 ) >= 2> from <rowSums(counts(dds) >= 10) >= 2>. This will increase the strength of the pre-filter and also filter low counts based on DESEq normalised values rather than absolute counts.
Also, REMOVED <lfcthreshold = 0.58> to <results(dds)>.'

cat(README, file = file.path(resultsDir, "README.txt"))
```

```{r}
metadata.csv <- paste0("metadata_", exp, ".csv")
# write_csv(metadata, file = file.path(resultsDir, metadata.csv))
```


```{r}
load("results/rnaseq/SRP104161/rawcounts/2024-06-04/rawcounts_786_OS_allcounts.RData")

# Extract the raw counts matrix from Rsubread output file and save to counts. 
counts <- feature_counts$counts

# Rename counts columns and metadata rows to match sample_name.
colnames(counts) <- sample_sheet$sample_name

# Subset counts matrix for samples to be assayed
counts <- counts[,c(metadata$sample_name)]

# Remove the"." version extension from the ensembl gene id's. The [.] specifies only remove after the first suffix.
rownames(counts) <- gsub("[.].*","",rownames(counts))
```


```{r}
# Define column metadata for DESeq2 object containing run number, ccRCC cell line and condition (meta vs nonmetastatic).
colData <- metadata %>%  
  dplyr::select(cell_line, 
         condition,
         batch)
        # Study)

# DESeq2 requires the metadata rownames match the count matrix column names. 
rownames(colData) <- metadata$sample_name

# Convert condition variables from characters to factors.
colData$condition <- as_factor(colData$condition)
colData$batch <- as_factor(colData$batch)
# Check the column names of the raw count matrix match the row names of metadata.
all(colnames(counts) == rownames(colData))
```

## Create DESeq2 Dataset

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                                  colData = colData,
                                  design = ~ condition) # batch + condition)

# Set reference factor for comparison
dds$condition <- relevel(dds$condition, ref = "nonmetastatic")
```

## Prefilter Raw Counts

```{r}
dds <- estimateSizeFactors(dds)
```

```{r}
dds <- estimateSizeFactors(dds)
smallestGroupSize <- 2
keep <- rowSums( counts(dds, normalized = TRUE) >= 10 ) >= smallestGroupSize
dds <- dds[keep,]
```

VST and rlog transform rawcounts (used for QC and plotting purposes, does not affect DEA)

```{r}
vsd <- vst(dds)
rld <- rlog(dds)
```

## Batch Correction (Likewise, used for QC and plotting, no effect on DEA)

```{r}
# Create a model matrix for the batch correction design formula (keep ~condition)
mm <- model.matrix(~condition, colData(vsd))
# Extract the matrix of transformed counts
vsd_mat <- assay(vsd)
# Remove batch effects with limma (batch already specified in metadata).
# Overwrite vsd matrix file. This is required for plotting only.
vsd_mat <- limma::removeBatchEffect(vsd_mat, batch=vsd$batch, design=mm)
assay(vsd) <- vsd_mat
```

## Differential Expression

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds, alpha = 0.05)#, lfcThreshold = 1) 
res
summary(res)
resultsNames(dds)
```

Generate the final results by applying log fold shrinkage to logfold change estimates using Apeglm method. This is recommended for reducing the contribution of low count genes to the final log fold change estimates. DESeq does NOT perform this step automatically. lfcShrink generates a new results table from the deseqdataset. Setting `res = res` ensures the alpha and lfcthreshold setting specified in the above results call are carried over to the shrunken results table.

```{r}
res_apeglm <- lfcShrink(dds, res = res ,coef="condition_metastatic_vs_nonmetastatic", type="apeglm")
res_apeglm

summary(res_apeglm)
```

Annotate results with Symbol IDs and descriptions ("GENENAME") from the org.Hs.eg.db database.

```{r}
# Extract ENSEMBLE IDs from the DESeq2 results object and use as mapping keys to retrieve ENTREZ and SYMBOL annotations.
ens.str <- c(rownames(res_apeglm)) #, 1, 15)
res_apeglm$symbol <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res_apeglm$genename <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
res_apeglm$biotype <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="GENETYPE",
                     keytype="ENSEMBL",
                     multiVals="first")
# Replace NA symbols with Ensembl IDs
ensembl_IDs <- rownames(res_apeglm)
symbols <- res_apeglm$symbol
symbols[is.na(symbols)] <- ensembl_IDs[is.na(symbols)]
res_apeglm$symbol <- symbols
```

Order results by padj and save

```{r}
res_ordered_by_padj <- res_apeglm[ order(res_apeglm$padj), ]

res.RData <- paste0("DESeq2_results_allTestedGenes_", exp, ".RData")
save(res_ordered_by_padj, file = file.path(resultsDir, res.RData))
```

```{r}
res_ordered_by_padj_df <- as.data.frame(res_ordered_by_padj) %>% 
  rownames_to_column(var = "geneID")

res.csv <- paste0("DESeq2_results_allTestedGenes_", exp, ".csv")
write.csv(res_ordered_by_padj_df, file = file.path(resultsDir, res.csv), row.names = FALSE)
```

By default the DESeq results object includes results for all genes tested. It is worth saving a copy of the complete results table as done above. The significant results should be manually subset using the same criteria applied when calling `results(dds)`. The `results` function actually applies the statistical test and generates the predicted p-values.

```{r}
res05 <- subset(res_ordered_by_padj, padj < 0.05)
res05

res05_Lfc05 <- subset(res05[abs(res05$log2FoldChange) > 0.58,])
res05_Lfc05

res_sig.RData <- paste0("DESeq2_results_FDR05_LFC58_", exp, ".RData")
save(res05_Lfc05, file = file.path(resultsDir, res_sig.RData))
```

```{r}
res05_lfc05_df <- as.data.frame(res05_Lfc05) %>% 
  rownames_to_column(var = "geneID")

res_sig.csv <- paste0("DESeq2_results_FDR05_LFC58_", exp, ".csv")
write.csv(res05_lfc05_df, file = file.path(resultsDir, res_sig.csv), row.names = FALSE)
```

```{r}
# Save DESeq normalised counts
norm_counts <- DESeq2::counts(dds, normalized = TRUE) %>% as.data.frame() %>% rownames_to_column(var = "gene_id")
file_name <- paste0("counts_deseq_normalised_786_OS.csv")
write.csv(norm_counts, file = file.path(resultsDir, file_name), col.names = TRUE, row.names = FALSE)
```


## Box Plot normalised vs non normalised raw counts

```{r}
boxplot_counts.pdf <- paste0("boxplot_norm_counts_", exp, ".pdf")
pdf(file = file.path(figureDir, boxplot_counts.pdf), width = 8)

par(mfrow=c(1,2)) # to plot the two box plots next to each other

print(boxplot(log2(counts(dds)), notch=TRUE,
main = "Non-normalized read counts\n(log-transformed)",
ylab="read counts"))

print(boxplot(log2(counts(dds, normalize= TRUE)), notch=TRUE,
main = "Size-factor-normalized read counts\n(log-transformed)",
ylab="read counts"))

dev.off()

boxplot_counts.svg <- paste0("boxplot_norm_counts_", exp, ".svg")
svg(file = file.path(figureDir, boxplot_counts.svg), width = 8)

par(mfrow=c(1,2)) # to plot the two box plots next to each other

print(boxplot(log2(counts(dds)), notch=TRUE,
main = "Non-normalized read counts\n(log-transformed)",
ylab="read counts"))

print(boxplot(log2(counts(dds, normalize= TRUE)), notch=TRUE,
main = "Size-factor-normalized read counts\n(log-transformed)",
ylab="read counts"))

dev.off()
```


## Scatterplots of transformed counts

Scatterplot of 786-O_rep1 against 786-M1A_rep1. (Note: check `colData(dds)` and select samples based on column order.

```{r}
df <- bind_rows(
  as.data.frame(log2(counts(dds, normalized=TRUE)[, c(1,3)]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as.data.frame(assay(vsd)[, c(1,3)]) %>% mutate(transformation = "vst"),
  as.data.frame(assay(rld)[, c(1,3)]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)

scatter_counts <- ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation) +
  theme_bw()

scatter_counts

scatter_counts.pdf <- paste0("scatter_transformed_counts_", exp, ".pdf")
pdf(file = file.path(figureDir, scatter_counts.pdf))
print(scatter_counts)
dev.off()

scatter_counts.svg <- paste0("scatter_transformed_counts_", exp, ".svg")
svg(file = file.path(figureDir, scatter_counts.svg))
print(scatter_counts)
dev.off()
```

## Sample to Sample Distances

Estimate distances between samples using the base R function `dist`

```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDists
```

Visualise sample to sample distances with pHeatmap

```{r}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$cell_line, vsd$batch, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Reds")) )(255)

samp_dists <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         #col = colors,
         cellwidth = 25,
         cellheight = 25,
         name = "Sample Distances")
         
samp_dists

samp_dist.pdf <- paste0("samp_dists_", exp, ".pdf")
pdf(file = file.path(figureDir, samp_dist.pdf), width = 8, height = 6)
print(samp_dists)
dev.off()

samp_dist.svg <- paste0("samp_dists_", exp, ".svg")
svg(file = file.path(figureDir, samp_dist.svg), width = 8, height = 6)
print(samp_dists)
dev.off()
```

## PCA

```{r}
pcaData <- plotPCA(vsd, intgroup = c("Study", "cell_line"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

pca_plot <- ggplot(pcaData, aes(x = PC1, y = PC2, color = cell_line, shape = Study)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw() +
  theme(aspect.ratio = 1) +
  scale_color_manual(values = c("786-O" = "#EE4B2B", "786-M1A" = "#26b3ff"))

pca_plot

pca.pdf <- paste0("PCA_corrected_", exp, ".pdf")
pdf(file = file.path(figureDir, pca.pdf), width = 6, height = 3)
print(pca_plot)
dev.off()

pca.svg <- paste0("PCA_corrected_", exp, ".svg")
svg(file = file.path(figureDir, pca.svg), width = 6, height = 3)
print(pca_plot)
dev.off()
```

## Expression Heatmap

Heatmap of all genes with FDR \< 0.05

```{r}
# Extract normalised counts
normalized_counts_df <- as.data.frame(counts(dds, normalized = TRUE))
# Subset the normalised counts to include only count data for row (ensgene) names present in the results data (padj < 0.1).
counts05 <- subset(normalized_counts_df[rownames(res05_Lfc05), ])

# Choose heatmap color palette
heat_colors <- brewer.pal(n = 6, name = "RdBu")
# Plot title
sig_genes_05 <- as.character(print(nrow(counts05)))
heat_title <- paste0(exp, " ", "[", sig_genes_05, " nDEGs FDR < 0.05 LFC > 0.58", "]")
# Plot heatmap of significant normalised counts
DE_heatmap <- pheatmap(counts05,
         color = heat_colors,
         cluster_rows = T, 
         show_rownames = F,
         annotation = dplyr::select(colData, condition), 
         scale = "row",  cutree_cols = 2,
         main = heat_title)

DE_heatmap

expression_heatmap.pdf <- paste0("DE_heatmap_sig_gene_", exp, ".pdf")
pdf(file = file.path(figureDir, expression_heatmap.pdf), width = 5, height = 5)
print(DE_heatmap)
dev.off()

expression_heatmap.svg <- paste0("DE_heatmap_sig_gene_", exp, ".svg")
svg(file = file.path(figureDir, expression_heatmap.svg), width = 5, height = 5)
print(DE_heatmap)
dev.off()
```

## MAplot

```{r}
MA_plot.pdf <- paste0("MAplot_apeglm_", exp, ".pdf")
pdf(file = file.path(figureDir, MA_plot.pdf), width = 4, height = 4)
plotMA(res_apeglm, ylim = c(-8, 8))
dev.off()

MA_plot.svg <- paste0("MAplot_apeglm_", exp, ".svg")
svg(file = file.path(figureDir, MA_plot.svg), width = 4, height = 4)
plotMA(res_apeglm, ylim = c(-8, 8))
dev.off()
```

## Volcano Plot

ggplot volcano V3

```{r}
# Set y-axis limit
y_axis_limit <- 100

# Create categorical column for up, down and .ns genes
ggplot_data <- res_ordered_by_padj_df
ggplot_data <- ggplot_data %>% 
  mutate(gene_type = case_when(log2FoldChange >= 1 & padj <= 0.05 ~ "up",
                               log2FoldChange <= -1 & padj <= 0.05 ~ "down",
                               TRUE ~ "ns"))

# Clip the -log10(padj) values to the y-axis limit
# ggplot_data <- ggplot_data %>% 
#   mutate(log10_padj_clipped = ifelse(-log10(padj) > y_axis_limit, y_axis_limit, -log10(padj)))

# Obtain gene counts
gene_counts <- ggplot_data %>% dplyr::count(gene_type)

up_count <- gene_counts %>% filter(gene_type == "up") %>% pull(n)
down_count <- gene_counts %>% filter(gene_type == "down") %>% pull(n)
ns_counts <- gene_counts %>% filter(gene_type == "ns") %>% pull(n)

# Add color, size, and transparency elements
cols <- c("up" = "red", "down" = "blue", "ns" = "grey") 
sizes <- c("up" = 2, "down" = 2, "ns" = 1) 
alphas <- c("up" = 1, "down" = 1, "ns" = 0.5)

# Genes to be labeled
genes_to_label <- c("CXCL8", "CXCL1", "CXCR4", "FZD8", "SOX9")

# Plot
ggVolcano <- ggplot(data = ggplot_data,
       aes(x = log2FoldChange,
           y = -log10(padj))) +
           #y = log10_padj_clipped)) + 
  geom_point(aes(colour = gene_type), 
             alpha = 0.4, 
             shape = 19,
             size = 2) + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(log2(0.5), log2(2)),
             linetype = "dashed") +
  scale_colour_manual(values = cols) + 
  scale_x_continuous(breaks = c(seq(-10, 10, 2)),     
                     limits = c(-10, 10)) +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols,
                      labels = c(
                       paste("Downregulated (", down_count, ")", sep = ""),
                       paste("Not Significant (", ns_counts, ")", sep = ""),
                       paste("Upregulated (", up_count, ")", sep = ""))) +
  scale_alpha_manual(values = alphas) + # Modify point transparency
  labs(x = "log2(fold change)",
       y = "-log10(adjusted P-value)",
       colour = "Expression \nchange") +
  theme_bw() + # Select theme with a white background  
  theme(panel.border = element_rect(colour = "black", fill = NA, size= 0.5)) +
  ylim(0, y_axis_limit) +
  geom_label_repel(data = subset(ggplot_data, symbol %in% genes_to_label),
                   aes(label = symbol),
                   size = 4,                # Increase label size
                   box.padding = 0.5,
                   point.padding = 0.2,
                   segment.color = "black",
                   segment.size = 0.5,
                   label.size = 0.25,       # Add a border to the label
                   label.padding = unit(0.2, "lines"),  # Padding inside label box
                   fill = "white")          # Fill color for label box

ggVolcano
volcano.pdf <- paste0("ggVolcano_Deseq2_apeglm_", exp, ".pdf")
pdf(file = file.path(figureDir, volcano.pdf), width = 6, height = 4)
print(ggVolcano)
dev.off()

volcano.svg <- paste0("ggVolcano_Deseq2_apeglm_", exp, ".svg")
svg(file = file.path(figureDir, volcano.svg), width = 6, height = 4)
print(ggVolcano)
dev.off()
```


## Pvalue Histogram

```{r}
Histogram_pVal.pdf <- paste0("Histogram_pValues_norm_counts_", exp, ".pdf")
pdf(file = file.path(figureDir, Histogram_pVal.pdf), width = 4, height = 4)
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()

Histogram_pVal.svg <- paste0("Histogram_pValues_norm_counts_", exp, ".svg")
svg(file = file.path(figureDir, Histogram_pVal.svg), width = 4, height = 4)
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()
```

## Heatmap of Top Differentially expressed Genes

Heatmap of top 10 genes (lfc up and down)

```{r}
# Subset sig genes to inlcude only those with base mean counts above 50 and absolute log2foldchange > 2
df.top <- as.data.frame(res05[ (res05$baseMean > 30) & (abs(res05$log2FoldChange) > 0.58),])
df.top

# Order by descending log2foldchange
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]

# Retrieve the batch corrected counts for df.top genes generated by limma earlier
vsd_out <- vst(dds, blind = FALSE) #Fetch normalised counts
mat <- assay(vsd_out)[rownames(df.top),] # Take sample names from dds coldata
colnames(mat) <- rownames(colData)
base_mean <- rowMeans(mat)
mat.scaled <- t(apply(mat, 1, scale)) #scale each column to 1 to get a z-score then transpose
colnames(mat.scaled) <- colnames(mat)
```

```{r}
# Select the number of genes to plot
num_keep <- 10
# Take the 10 top and 10 bottom genes from the matrix (that we previously ordered by lfc = decreasing)
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled) - num_keep), nrow(mat.scaled)))

# Get the log2 fold change values for genes to keep
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange)
colnames(l2_val) <- "log2FC"

mean <- as.matrix(df.top[rows_keep,]$baseMean)
colnames(mean) <- "AveExpr"
```

Plot the heatmap whoop!

```{r}
# prepare some colours for the heatmap
# maps colours between blue white and red for l2fc value range
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))
# maps colour to 4 quantiles of aveExpr vals
col_aveExr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))
```

```{r}
ha <- HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2),
                                               height = unit(2, "cm")))

h1 <- Heatmap(mat.scaled[rows_keep,], cluster_rows = F,
              column_labels = colnames(mat.scaled), name = "Z score",
              cluster_columns = T, column_names_rot = 90)
h2 <- Heatmap(l2_val, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "Log2FC", #top_annotation = ha, 
              col = col_logFC,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(l2_val[i, j], 2), x, y)
                })
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "AveExpr", top_annotation = ha, col = col_aveExr,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(mean[i, j], 2), x, y)
                })
h <- h1+h2
h

top10_heattitle.svg <- paste0("Top10DEGenes_Heatmap_", exp, ".svg")
svg(file = file.path(figureDir, top10_heattitle.svg), width = 8, height = 10)
print(h)
dev.off()

top10_heattitle.pdf <- paste0("Top10DEGenes_Heatmap_", exp, ".pdf")
pdf(file = file.path(figureDir, top10_heattitle.pdf), width = 8, height = 10)
print(h)
dev.off()
```

Heatmap for top 25 genes both up and down

```{r}
# Select the number of genes to plot
num_keep <- 25
# Take the 10 top and 10 bottom genes from the matrix (that we previously ordered by lfc = decreasing)
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled) - num_keep), nrow(mat.scaled)))

# Get the log2 fold change values for genes to keep
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange)
colnames(l2_val) <- "log2FC"

mean <- as.matrix(df.top[rows_keep,]$baseMean)
colnames(mean) <- "AveExpr"
```

Plot the heatmap whoop!

```{r}
# prepare some colours for the heatmap
# maps colours between blue white and red for l2fc value range
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))
# maps colour to 4 quantiles of aveExpr vals
col_aveExr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))
```

```{r}
ha <- HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2),
                                               height = unit(2, "cm")))

h1 <- Heatmap(mat.scaled[rows_keep,], cluster_rows = F,
              column_labels = colnames(mat.scaled), name = "Z score",
              cluster_columns = T, column_names_rot = 90)
h2 <- Heatmap(l2_val, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "Log2FC", #top_annotation = ha, 
              col = col_logFC,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(l2_val[i, j], 2), x, y)
                })
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "AveExpr", top_annotation = ha, col = col_aveExr,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(mean[i, j], 2), x, y)
                })
h <- h1+h2
h

top25_heattitle.svg <- paste0("Top25DEGenes_Heatmap_", exp, ".svg")
svg(file = file.path(figureDir, top25_heattitle.svg), width = 8, height = 14)
print(h)
dev.off()

top25_heattitle.pdf <- paste0("Top25DEGenes_Heatmap_", exp, ".pdf")
pdf(file = file.path(figureDir, top25_heattitle.pdf), width = 8, height = 14)
print(h)
dev.off()
```

Heatmap for top 50 genes both up and down

```{r}
# Select the number of genes to plot
num_keep <- 50
# Take the 10 top and 10 bottom genes from the matrix (that we previously ordered by lfc = decreasing)
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled) - num_keep), nrow(mat.scaled)))

# Get the log2 fold change values for genes to keep
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange)
colnames(l2_val) <- "log2FC"

mean <- as.matrix(df.top[rows_keep,]$baseMean)
colnames(mean) <- "AveExpr"
```

Plot the heatmap whoop!

```{r}
# prepare some colours for the heatmap
# maps colours between blue white and red for l2fc value range
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))
# maps colour to 4 quantiles of aveExpr vals
col_aveExr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))
```

```{r}
ha <- HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2),
                                               height = unit(2, "cm")))

h1 <- Heatmap(mat.scaled[rows_keep,], cluster_rows = F,
              column_labels = colnames(mat.scaled), name = "Z score",
              cluster_columns = T, column_names_rot = 90)
h2 <- Heatmap(l2_val, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "Log2FC", #top_annotation = ha, 
              col = col_logFC,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(l2_val[i, j], 2), x, y)
                })
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep],
              cluster_rows = F, name = "AveExpr", top_annotation = ha, col = col_aveExr,
              cell_fun = function(j, i, x, y, w, h, col){ #This madness adds anno text to the anno columns
                grid.text(round(mean[i, j], 2), x, y)
                })
h <- h1+h2
h

top50_heattitle.svg <- paste0("Top50DEGenes_Heatmap_", exp, ".svg")
svg(file = file.path(figureDir, top50_heattitle.svg), width = 8, height = 30)
print(h)
dev.off()

top50_heattitle.pdf <- paste0("Top50DEGenes_Heatmap_", exp, ".pdf")
pdf(file = file.path(figureDir, top50_heattitle.pdf), width = 8, height = 30)
print(h)
dev.off()
```


## Quick <GO:BP> with ClusterProfiler

```{r}
upsigs <- rownames(res05_Lfc05[ (res05_Lfc05$baseMean > 30) & (res05_Lfc05$log2FoldChange) >= 1,])
downsigs <- rownames(res05_Lfc05[ (res05_Lfc05$baseMean > 30) & (res05_Lfc05$log2FoldChange) <= -1,])

universe <- rownames(res_apeglm)
```

All DE genes

```{r}
gobp <- enrichGO(gene = res05_lfc05_df$geneID,
                universe = universe,
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                keyType = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
head(gobp)

upgo.df <- as.data.frame(gobp)
upgo.name <- paste0("clusterprofiler_results_GO_BP_ALLSIGGENES_", exp, ".tsv")
write.csv(upgo.df, file = file.path(resultsDir, upgo.name), row.names = FALSE)
```

```{r}
# Remove redundant go terms
gobp$Description
gobp_select <- gobp$Description[c(2,3,5,6,8,11,15,16,17,19)]

upgo.dot <- barplot(gobp, showCategory=gobp_select) + ggtitle("Biological Process Enrichment")

UPBP_title.svg <- paste0("TopGOBP_ALLSIGGENES_", exp, ".svg")
svg(file = file.path(figureDir, UPBP_title.svg), width = 5, height = 6)
print(upgo.dot)
dev.off()

UPBP_title.pdf <- paste0("TopGOBP_ALLSIGGENES_", exp, ".pdf")
pdf(file = file.path(figureDir, UPBP_title.pdf), width = 5, height = 6)
print(upgo.dot)
dev.off()
```

Up and Down genes

```{r}
upgo <- enrichGO(gene = upsigs,
                universe = universe,
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                keyType = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
head(upgo)

upgo.df <- as.data.frame(upgo)
upgo.name <- paste0("clusterprofiler_results_UPGO_BP_", exp, ".tsv")
write.csv(upgo.df, file = file.path(resultsDir, upgo.name), row.names = FALSE)
```

```{r}
downgo <- enrichGO(gene = downsigs,
                universe = universe,
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                keyType = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
head(downgo)

downgo.df <- as.data.frame(downgo)
downgo.name <- paste0("clusterprofiler_results_DOWNGO_BP_", exp, ".tsv")
write.csv(downgo.df, file = file.path(resultsDir, downgo.name), row.names = FALSE)
```

```{r}
upgo.dot <- dotplot(upgo, showCategory=10) + ggtitle("Postive Biological Process Enrichment")

UPBP_title.svg <- paste0("TopUPGOBP_", exp, ".svg")
svg(file = file.path(figureDir, UPBP_title.svg), width = 7, height = 6)
print(upgo.dot)
dev.off()

UPBP_title.pdf <- paste0("TopUPGOBP_", exp, ".pdf")
pdf(file = file.path(figureDir, UPBP_title.pdf), width = 7, height = 6)
print(upgo.dot)
dev.off()
```

```{r}
downgo.dot <- dotplot(downgo, showCategory=10) + ggtitle("Negative Enrichment of Biological Processes")

DOWNBP_title.svg <- paste0("TopDOWNGOBP_", exp, ".svg")
svg(file = file.path(figureDir, DOWNBP_title.svg), width = 7, height = 6)
print(downgo.dot)
dev.off()

DOWNBP_title.pdf <- paste0("TopDOWNGOBP_", exp, ".pdf")
pdf(file = file.path(figureDir, DOWNBP_title.pdf), width = 7, height = 6)
print(downgo.dot)
dev.off()
```

## Comparison of Deseq Results for the different cell lines using venn diagrams of overlapping ENSEMBL gene IDs

```{r}
res_786 <- read.csv(file = file.path(resultsDir, "DESeq2_results_FDR05_LFC58_786_vs_M1A.csv"))
res_786_batch <- read.csv(file = file.path(resultsDir, "DESeq2_results_FDR05_LFC58_786O_vs_M1A_batch.csv"))

genes <- list(Rodrigues_et_el = res_786$geneID,
              Sciacovelli_et_al = res_786_batch$geneID)
venn.diagram(genes, filename = "786O_M1A_batch")
```

```{r}
res_786 <- read.csv(file = file.path(resultsDir, "DESeq2_results_FDR05_LFC58_786_vs_M1A.csv"))

res_OS <- read.csv(file = file.path(resultsDir, "DESeq2_results_FDR05_LFC58_OS-RC2_LM1.csv"))

res_786_sig <- res_786 %>% dplyr::filter(log2FoldChange >= 1 | log2FoldChange <= -1)
res_OS_sig <- res_OS %>% dplyr::filter(log2FoldChange >= 1 | log2FoldChange <= -1)

vennlist <- list(`786-O vs M1A` = res_786_sig$geneID,
                 `OS-RC2 vs LM1` = res_OS_sig$geneID)


v1 <- ggvenn(vennlist, 
             fill_color = c("blue", "red"),
             fill_alpha = 0.7,
             text_color = "black",
             text_size = 6)

file_name <- paste0("ggvenn_DEG_overlap_786vsOS.pdf")
pdf(file = file.path(figureDir, file_name))
print(v1)
dev.off()

file_name <- paste0("ggvenn_DEG_overlap_786vsOS.svg")
svg(file = file.path(figureDir, file_name))
print(v1)
dev.off()

common_DEGs <- inner_join(res_786_sig, res_OS_sig, by = "geneID")
file_name <- paste0("DESeq2_results_commonDEGs_786vsOS.csv")
write.csv(common_DEGs, file = file.path(resultsDir, file_name), row.names = FALSE)
```

```{r}
go <- enrichGO(gene = common_DEGs$geneID,
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                keyType = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)
head(go)

go.df <- as.data.frame(go)
go.name <- paste0("clusterprofiler_results_GO_BP_commonDEGs_786vsOS.tsv")
write.csv(go.df, file = file.path(resultsDir, go.name), row.names = FALSE)


go.bar <- barplot(go, showCategory=10) + ggtitle("Enrichment of Biological Processes")
go.bar

BP_title.svg <- paste0("GOBP_commonDEGs_786vsOS_wide.svg")
svg(file = file.path(figureDir, BP_title.svg), width = 8, height = 5)
print(go.bar)
dev.off()

BP_title.pdf <- paste0("GOBP_commonDEGs_786vsOS_wide.pdf")
pdf(file = file.path(figureDir, BP_title.pdf), width = 8, height = 5)
print(go.bar)
dev.off()
```

