#!/usr/bin/bash

#PBS -N deeptoolscomputeMatrix
#PBS -l walltime=24:00:00
#PBS -l select=1:ncpus=8:mem=64gb
#PBS -J 1-4
#Sets current directory to working
cd ${PBS_O_WORKDIR}

module load anaconda3/personal
source activate deeptools

# SETUP FOR PBS ARRAY
# COMMENT OUT FOR SINGLE BATCH SUBMISSIONS
#####################################################################################
## Setup PBS array and assign file names stored in metadata using PBS_ARRAY_INDEX
#####################################################################################
metadata=$HOME/ThesisProject/metadata/chipseq/SRP104404/metadata_computeMatrix.csv
chip1=$(awk -F ',' -v taskID=${PBS_ARRAY_INDEX} '$1==taskID {print $2}' ${metadata})
chip2=$(awk -F ',' -v taskID=${PBS_ARRAY_INDEX} '$1==taskID {print $3}' ${metadata})

lab1=$(awk -F ',' -v taskID=${PBS_ARRAY_INDEX} '$1==taskID {print $4}' ${metadata})
lab2=$(awk -F ',' -v taskID=${PBS_ARRAY_INDEX} '$1==taskID {print $5}' ${metadata})

exp=$(awk -F ',' -v taskID=${PBS_ARRAY_INDEX} '$1==taskID {print $6}' ${metadata})
#####################################################################################

# File extensions
bw=_chip_merged.bam.bw

# Create inpath and outpaths
inpath=$HOME/data/SRP104404/bigwig/merge
outpath=$HOME/ThesisProject/results/chipseq/SRP104404/deeptools/computeMatrix
output=${outpath}/computeMatrix_${exp}

#####################################################################################
# input files
#####################################################################################
bw1=${inpath}/${chip1}${bw}
bw2=${inpath}/${chip2}${bw}

gtf=$HOME/data/reference_genome/gencode/GRCh38/gencode.v45.primary_assembly.annotation.gtf

# creates outpath if it doesn't exist
mkdir -p ${outpath}

computeMatrix reference-point \
--referencePoint TSS \
 -b 2500 -a 2500 \
 -R ${gtf} \
 -S ${bw1} ${bw2} \
--samplesLabel ${lab1} ${lab2} \
--skipZeros \
 -o ${output}.gz \
--outFileSortedRegions ${output}_gencodeTSSregions.bed
